pkg/iss/iss.go

```diff
@@ -703,6 +705,57 @@ func (iss *ISS) applyCheckpointMessage(message *isspb.Checkpoint, source t.NodeI
 	}
 }
 
+// applyStableCheckpointMessage applies a StableCheckpoint message
+// received over the network. It verifies the message and decides
+// whether to install the state snapshot from the message.
+func (iss *ISS) applyStableCheckpointMessage(m *isspb.StableCheckpoint, source t.NodeID) *events.EventList {
+
+	// Check how far is the received stable checkpoint ahead of
+	// the local node's state.
+	if t.EpochNr(m.Epoch) <= iss.epoch.Nr+1 {
+		// Ignore stable checkpoints that are not far enough
+		// ahead of the current state of the local node.
+		return &events.EventList{}
+	}
+
+	iss.logger.Log(logging.LevelDebug, "Installing state snapshot.", "epoch", m.Epoch)
+
+	// Clean up global ISS state that belongs to the current epoch
+	// instance that local replica got stuck with.
+	iss.epochs = make(map[t.EpochNr]*epochInfo)
+	iss.epoch = nil
+	iss.commitLog = make(map[t.SeqNr]*CommitLogEntry)
+	iss.unhashedLogEntries = make(map[t.SeqNr]*CommitLogEntry)
+	iss.nextDeliveredSN = t.SeqNr(m.Sn)
+	iss.newEpochSN = iss.nextDeliveredSN
+
+	// Initialize a new ISS epoch instance for the new stable
+	// checkpoint to continue participating in the protocol
+	// starting with that epoch after installing the state
+	// snapshot from the new stable checkpoint.
+	iss.initEpoch(t.EpochNr(m.Epoch))
+
+	// Update the last stable checkpoint stored in the global ISS structure.
+	iss.lastStableCheckpoint = m
+
+	// Create an event to request the application module for
+	// restoring its state from the snapshot received in the new
+	// stable checkpoint message.
+	restoreEvent := events.AppRestoreState(m.AppSnapshot)
+
+	// Attach a follow-up event that activates SB instances of the
+	// new epoch after the application module has restored the
+	// state from the snapshot.
+	restoreEvent.FollowUps(iss.initOrderers().Slice())
+
+	// Attach a follow-up event that applies any message buffered
+	// for the new epoch after the application module has restored
+	// the state from the snapshot.
+	restoreEvent.FollowUps(iss.applyBufferedMessages().Slice())
```

Comment:
Do these need to be follow-up events?
It seems to me that maybe the events could just be returned in a list.
Events in a single list that are destined to the same module are guaranteed to be applied in FIFO order to that module.

Reply:
You seem to be right, the new SB instances would't be able to interact with the application module anyway until the state restore event is fully handled, so there's apparently no need to delay activation of the new epoch instance.

Reply:
However, what if the new epoch instance happens to deliver a new batch (e.g. while processing the buffered events) before the restore state event reaches the application module's event queue?

Reply:
That doesn't seem possible to me, since it would have to happen as a direct result of the application of the `StableCheckpoint` message and precede the `AppRestoreState` event in the returned event list. By the time any other event is processed by the protocol, all the events resulting from the application of `StableCheckpoint` are already queued up.

Reply:
> By the time any other event is processed by the protocol, all the events resulting from the application of StableCheckpoint are already queued up.

Is it guaranteed that all events returned by a module always get queued to the destination event queues in order, even those that are returned as a result of processing by that module subsequent incoming events?

Reply:
This seems to be the case as long as `Node.workChans.workItemInput` channel is not buffered.

Reply:
I think it would even work if the channel was buffered. Outgoing events from a single module (in this case the `Protocol` module) are never reordered with respect to each other. A single thread injects the events in the `Protocol` module and the very same thread collects the resulting events. Thus, any result of processing the `StableCheckpoint` message (technically the `MessageReceived` event containing the `StableCheckpoint`) will be written to `workItemInput` (buffered or not) before any result of processing subsequent events, as they are handled by the same thread as the `StableCheckpoint`.

Reply:
Note that this is not necessarily true for follow-up events. A follow-up event is only guaranteed to be processed after the event it is attached to, but not necessarily immediately after it. Arbitrary events, even originating in the same module, may pass in between.